StrongMind/canvas_shim
# QTY-1355: Add assignment group passing thresholds to account settings

[QTY-1355](https://strongmind.atlassian.net/browse/QTY-1355)

## Purpose 

Allow school personnel to set a school wide default minimum threshold at the account level that is reflected on all new and currently created (but not yet started) courses by assignment group. 

## Approach 

Update account settings page to have fields for a pre-determined set of assignment groups, replacing the existing threshold fields for assignments and exams. 

Adds `AssignmentGroup::GROUP_NAMES` constant to hold the assignment groups for which we will support adding default thresholds. 

Utilizes the existing `RequirementsService` and `SettingsService` to dynamically get and set thresholds for the supported assignment groups. 

Adds callbacks to `assignment_decorator.rb` and `content_migration_decorator.rb` as well as delayed jobs in `account_decorator.rb`, `context_module_decorator.rb`, and `course_decorator.rb` to handle the various ways in which an assignment group might change and update the threshold appropriately. 

## Testing

Tested locally and on newidsandbox. To confirm thresholds are set and updated as expected some example steps might be: 

1. Set thresholds on account settings page
2. Create a new course and import a cartridge, verify that the thresholds set are applied based on the content's assignment group
3. Edit an assignments assignment group, verify that the threshold changed as appropriate
4. Manually create an assignment and a discussion topic (since they are created differently) and verify that both have the appropriate thresholds set (DiscussionTopic will not have a threshold set until marked as `graded` as that is when it has an associated assignment created and therefor an assignment group threshold to associate)
5. Create courses with a start date in the past and one in the future
6. Change the thresholds set in account settings
7. Verify that the course with a past start date is not affected but the course with the future start date reflects the new thresholds

## Screenshots

#### Before: 
<img width="1218" alt="image" src="https://user-images.githubusercontent.com/30609917/206741028-8cd4604a-0c9c-48f5-ac4b-26d77da5e42a.png">


#### After: 
<img width="1268" alt="image" src="https://user-images.githubusercontent.com/30609917/206740925-bd22b559-2067-4ed4-8c8b-f591ef8d657e.png">
